File: src/bin/run_onnx.rs
use color_eyre::Result;
use image::{DynamicImage, GenericImageView, ImageBuffer, Rgb, imageops::FilterType};
use ndarray::{Array1, Array2, Array4, Axis, s};
use ort::session::Session;
use ort::value::Value;
use serde_json;
use std::fs;
use std::path::Path;

#[derive(Debug, Clone)]
pub struct Detection {
    pub bbox: [f32; 4],
    pub score: f32,
    pub class_id: usize,
    pub tag: String,
    pub mask: Option<Array2<bool>>,
}

pub struct YOLO26Predictor {
    session: Session,
    vocab: Vec<String>,
    imgsz: u32,
    stride: u32,
}

impl YOLO26Predictor {
    pub fn new(model_path: impl AsRef<Path>, vocab_path: impl AsRef<Path>) -> Result<Self> {
        let session = Session::builder()?
            .with_optimization_level(ort::session::builder::GraphOptimizationLevel::Level3)
            .expect("with_optimization_level")
            .with_intra_threads(num_cpus::get())
            .expect("with_intra_threads")
            .commit_from_file(model_path)?;

        let vocab_file = fs::read_to_string(vocab_path)?;
        let vocab: Vec<String> = serde_json::from_str(&vocab_file)?;

        Ok(Self {
            session,
            vocab,
            imgsz: 640,
            stride: 32,
        })
    }

    /// Replicates the Python standalone letterbox logic exactly
    fn preprocess(&self, img: &DynamicImage, debug_name: Option<&str>) -> Result<(Array4<f32>, PreprocessMeta)> {
        let (w0, h0) = img.dimensions();

        // 1. Calculate scaling ratio (long side = 640)
        let r = self.imgsz as f32 / (w0.max(h0) as f32);

        // 2. Calculate new unpadded dimensions (matches Python's int(round(w0 * r)))
        let new_unpad_w = (w0 as f32 * r).round() as u32;
        let new_unpad_h = (h0 as f32 * r).round() as u32;

        // 3. Calculate rectangular padding (nearest multiple of stride)
        let w_pad = ((new_unpad_w as f32 / self.stride as f32).ceil() * self.stride as f32) as u32;
        let h_pad = ((new_unpad_h as f32 / self.stride as f32).ceil() * self.stride as f32) as u32;

        let resized = img.resize_exact(new_unpad_w, new_unpad_h, FilterType::CatmullRom);

        // 5. Create Gray Canvas (114)
        let mut canvas = ImageBuffer::from_pixel(w_pad, h_pad, Rgb([114, 114, 114]));

        // Python: top, left = dh // 2, dw // 2
        let dw = w_pad - new_unpad_w;
        let dh = h_pad - new_unpad_h;
        let left = dw / 2;
        let top = dh / 2;

        image::imageops::overlay(&mut canvas, &resized.to_rgb8(), left as i64, top as i64);

        // 6. Normalize and Convert to CHW
        // Important: Python's BGR-to-RGB conversion is handled by Rust's DynamicImage::open (which is RGB)
        let mut input = Array4::zeros((1, 3, h_pad as usize, w_pad as usize));
        for (x, y, rgb) in canvas.enumerate_pixels() {
            input[[0, 0, y as usize, x as usize]] = (rgb[0] as f32) / 255.0;
            input[[0, 1, y as usize, x as usize]] = (rgb[1] as f32) / 255.0;
            input[[0, 2, y as usize, x as usize]] = (rgb[2] as f32) / 255.0;
        }

        // --- DEBUG VISUALIZATION ---
        if let Some(name) = debug_name {
            let mut debug_img = ImageBuffer::new(w_pad, h_pad);
            for y in 0..h_pad {
                for x in 0..w_pad {
                    let r = (input[[0, 0, y as usize, x as usize]] * 255.0) as u8;
                    let g = (input[[0, 1, y as usize, x as usize]] * 255.0) as u8;
                    let b = (input[[0, 2, y as usize, x as usize]] * 255.0) as u8;
                    debug_img.put_pixel(x, y, Rgb([r, g, b]));
                }
            }
            debug_img.save(format!("rust_debug_{}", name))?;
        }
        // ---------------------------

        let meta = PreprocessMeta {
            ratio: r,
            pad: (left as f32, top as f32),
            orig_shape: (w0, h0),
            tensor_shape: (w_pad, h_pad),
        };

        Ok((input, meta))
    }

    pub fn predict(&mut self, img_path: impl AsRef<Path>, conf_threshold: f32, iou_threshold: f32) -> Result<Vec<Detection>> {
        let file_name = img_path.as_ref().file_name().unwrap().to_str().unwrap();
        let img = image::open(&img_path)?;

        // Pass the filename to save debug image
        let (input_tensor, meta) = self.preprocess(&img, Some(file_name))?;

        let input_value = Value::from_array(input_tensor)?;
        let outputs = self.session.run(ort::inputs!["images" => input_value])?;

        let detections_owned = outputs["detections"].try_extract_array::<f32>()?.to_owned();
        let protos_owned = outputs["protos"].try_extract_array::<f32>()?.to_owned();
        drop(outputs);

        let preds = detections_owned.slice(s![0, .., ..]);
        let protos = protos_owned.slice(s![0, .., .., ..]);

        let mut candidates = Vec::new();
        for i in 0..preds.shape()[0] {
            let score = preds[[i, 4]];
            if score > conf_threshold {
                let class_id = preds[[i, 5]] as usize;
                let bbox = [preds[[i, 0]], preds[[i, 1]], preds[[i, 2]], preds[[i, 3]]];
                let mask_weights = preds.slice(s![i, 6..38]).to_owned();
                candidates.push((bbox, score, class_id, mask_weights));
            }
        }

        let kept_indices = Self::nms(&candidates, iou_threshold);

        let mut results = Vec::new();
        for idx in kept_indices {
            let (bbox, score, class_id, weights) = &candidates[idx];

            let x1 = (bbox[0] - meta.pad.0) / meta.ratio;
            let y1 = (bbox[1] - meta.pad.1) / meta.ratio;
            let x2 = (bbox[2] - meta.pad.0) / meta.ratio;
            let y2 = (bbox[3] - meta.pad.1) / meta.ratio;

            let mask = Self::process_mask(&protos, weights, &meta);

            results.push(Detection {
                bbox: [x1, y1, x2, y2],
                score: *score,
                class_id: *class_id,
                tag: self.vocab.get(*class_id).cloned().unwrap_or_else(|| "unknown".to_string()),
                mask: Some(mask),
            });
        }

        Ok(results)
    }

    // Helper changed to associated function (no &self)
    fn nms(
        candidates: &Vec<([f32; 4], f32, usize, Array1<f32>)>,
        iou_threshold: f32,
    ) -> Vec<usize> {
        if candidates.is_empty() {
            return vec![];
        }
        let mut indices: Vec<usize> = (0..candidates.len()).collect();
        indices.sort_by(|&a, &b| candidates[b].1.partial_cmp(&candidates[a].1).unwrap());

        let mut kept = Vec::new();
        while let Some(current) = indices.first().cloned() {
            kept.push(current);
            indices.remove(0);
            indices.retain(|&idx| {
                Self::calculate_iou(&candidates[current].0, &candidates[idx].0) <= iou_threshold
            });
        }
        kept
    }

    fn calculate_iou(box1: &[f32; 4], box2: &[f32; 4]) -> f32 {
        let x1 = box1[0].max(box2[0]);
        let y1 = box1[1].max(box2[1]);
        let x2 = box1[2].min(box2[2]);
        let y2 = box1[3].min(box2[3]);
        let intersection = (x2 - x1).max(0.0) * (y2 - y1).max(0.0);
        let area1 = (box1[2] - box1[0]) * (box1[3] - box1[1]);
        let area2 = (box2[2] - box2[0]) * (box2[3] - box2[1]);
        intersection / (area1 + area2 - intersection)
    }

    // Helper changed to associated function (no &self)
    fn process_mask(
        protos: &ndarray::ArrayView3<f32>,
        weights: &Array1<f32>,
        meta: &PreprocessMeta,
    ) -> Array2<bool> {
        let (c_proto, h_proto, w_proto) = protos.dim();
        let mut mask_flat = Array2::<f32>::zeros((h_proto, w_proto));
        for i in 0..c_proto {
            let w = weights[i];
            let p_layer = protos.index_axis(Axis(0), i);
            mask_flat += &(p_layer.to_owned() * w);
        }

        let mask_sigmoid = mask_flat.mapv(|x| 1.0 / (1.0 + (-x).exp()));

        let mask_img = ImageBuffer::from_fn(w_proto as u32, h_proto as u32, |x, y| {
            image::Luma([(mask_sigmoid[[y as usize, x as usize]] * 255.0) as u8])
        });

        let dyn_mask = DynamicImage::ImageLuma8(mask_img);
        let upscaled = dyn_mask.resize_exact(
            meta.tensor_shape.0,
            meta.tensor_shape.1,
            FilterType::Triangle,
        );

        let (w_ori, h_ori) = meta.orig_shape;
        let crop_w = (w_ori as f32 * meta.ratio).round() as u32;
        let crop_h = (h_ori as f32 * meta.ratio).round() as u32;
        let cropped = upscaled.crop_imm(meta.pad.0 as u32, meta.pad.1 as u32, crop_w, crop_h);

        let final_mask = cropped
            .resize_exact(w_ori, h_ori, FilterType::Triangle)
            .to_luma8();

        Array2::from_shape_fn((h_ori as usize, w_ori as usize), |(y, x)| {
            final_mask.get_pixel(x as u32, y as u32)[0] > 127
        })
    }
}

struct PreprocessMeta {
    ratio: f32,
    pad: (f32, f32),
    orig_shape: (u32, u32),
    tensor_shape: (u32, u32),
}

pub fn main() -> Result<()> {
    let model_path = "assets/model/dynamic-onnx/yoloe-26l-seg-pf-dynamic-try-3.onnx";
    let vocab_path = "assets/model/dynamic-onnx/vocabulary-dynamic.json";

    let mut predictor = YOLO26Predictor::new(model_path, vocab_path)?;

    println!("--- YOLO26 Rust Standalone Inference ---");

    let img_dir = Path::new("assets/img");
    for entry in fs::read_dir(img_dir)? {
        let path = entry?.path();
        if let Some(ext) = path.extension() {
            if ext == "jpg" || ext == "jpeg" || ext == "png" {
                let start = std::time::Instant::now();
                let results = predictor.predict(&path, 0.4, 0.7)?;
                let duration = start.elapsed();

                println!("Image: {:?} ({:?})", path.file_name().unwrap(), duration);
                println!("  - Objects Found: {}", results.len());

                let tags: std::collections::HashSet<_> = results.iter().map(|r| &r.tag).collect();
                if !tags.is_empty() {
                    println!("  - Tags: {:?}", tags);
                }
                println!("------------------------------");
            }
        }
    }

    Ok(())
}

File: py-yolo/run_onnx.py
# /// script
# requires-python = "==3.12.*"
# dependencies = [
# "onnx==1.21.0",
# "opencv-python==4.13.0.92",
# "onnxruntime==1.24.4",
# ]
# ///
import cv2
import numpy as np
import onnxruntime as ort
import json
import time
from pathlib import Path


class YOLO26StandalonePredictor:
    def __init__(self, onnx_path, vocab_path):
        # Load ONNX Session
        self.session = ort.InferenceSession(onnx_path, providers=['CPUExecutionProvider'])

        # Load Vocabulary
        with open(vocab_path, 'r') as f:
            self.vocab = json.load(f)

        self.imgsz = 640
        self.stride = 32

    def preprocess(self, img0):
        """
        Manual implementation of Ultralytics Letterbox (rect=True, auto=False).
        """
        h0, w0 = img0.shape[:2]

        # 1. Calculate scaling ratio (long side = 640)
        r = self.imgsz / max(h0, w0)

        # 2. Calculate new unpadded dimensions
        new_unpad_w = int(round(w0 * r))
        new_unpad_h = int(round(h0 * r))

        # 3. Calculate rectangular padding (nearest multiple of stride)
        # This is exactly what the 'rect' diagnostic showed (e.g., 384x640)
        h_pad = int(np.ceil(new_unpad_h / self.stride) * self.stride)
        w_pad = int(np.ceil(new_unpad_w / self.stride) * self.stride)

        # 4. Resize
        img = cv2.resize(img0, (new_unpad_w, new_unpad_h), interpolation=cv2.INTER_LINEAR)

        # 5. Calculate padding for centering
        dw = w_pad - new_unpad_w
        dh = h_pad - new_unpad_h

        top, bottom = dh // 2, dh - (dh // 2)
        left, right = dw // 2, dw - (dw // 2)

        # 6. Apply Padding (Gray color: 114)
        img = cv2.copyMakeBorder(img, top, bottom, left, right, cv2.BORDER_CONSTANT, value=(114, 114, 114))

        # 7. BGR to RGB, HWC to CHW, Normalize to 0-1
        img = img.transpose((2, 0, 1))[::-1]  # BGR to RGB, CHW
        img = np.ascontiguousarray(img).astype(np.float32) / 255.0

        # Store metadata for rescaling
        meta = {
            "ratio": r,
            "pad": (left, top),
            "orig_shape": (h0, w0),
            "tensor_shape": (h_pad, w_pad)
        }

        return img[None], meta

    def process_masks(self, protos, mask_weights, meta):
        """Standalone mask reconstruction logic."""
        if len(mask_weights) == 0: return []

        h_ori, w_ori = meta["orig_shape"]
        h_new, w_new = meta["tensor_shape"]
        left, top = meta["pad"]
        r = meta["ratio"]

        mh, mw = protos.shape[1:]  # 160x160

        # Matrix multiplication of weights and protos
        masks = (mask_weights @ protos.reshape(32, -1)).reshape(-1, mh, mw)
        masks = 1 / (1 + np.exp(-masks))  # Sigmoid

        results = []
        for mask in masks:
            # Upscale mask to the network input size (e.g. 384x640)
            mask = cv2.resize(mask, (w_new, h_new), interpolation=cv2.INTER_LINEAR)
            # Crop out the gray padding
            mask = mask[top:top + int(h_ori * r), left:left + int(w_ori * r)]
            # Scale back to original image pixels
            mask = cv2.resize(mask, (w_ori, h_ori), interpolation=cv2.INTER_LINEAR)
            results.append(mask > 0.5)
        return results

    def predict(self, img_path, conf=0.4, iou=0.7):
        img0 = cv2.imread(str(img_path))
        if img0 is None: return [], 0, 0

        # 1. Preprocess
        img_input, meta = self.preprocess(img0)
        # --- DEBUG: VISUALIZE STANDALONE PREPROCESSING ---
        # 1. Take the first image in batch [1, 3, H, W] -> [3, H, W]
        # 2. Transpose to [H, W, 3]
        debug_vis = img_input[0].transpose(1, 2, 0)
        # 3. Denormalize 0-1 to 0-255
        debug_vis = (debug_vis * 255).astype(np.uint8)
        # 4. Convert RGB (internal) back to BGR (OpenCV)
        debug_vis = cv2.cvtColor(debug_vis, cv2.COLOR_RGB2BGR)
        # 5. Save the image
        cv2.imwrite(f"standalone_debug_{Path(img_path).name}", debug_vis)
        # -------------------------------------------------

        # --- EXPORT FOR RUST VERIFICATION ---
        debug_dir = Path("debug_data")
        debug_dir.mkdir(exist_ok=True)

        stem = Path(img_path).stem
        # Save exact float32 binary
        img_input.astype(np.float32).tofile(debug_dir / f"{stem}_tensor.bin")

        # Save exact metadata
        with open(debug_dir / f"{stem}_meta.json", "w") as f:
            json.dump({
                "shape": img_input.shape,  # [1, 3, H, W]
                "ratio": meta["ratio"],
                "pad": meta["pad"],
                "orig_shape": meta["orig_shape"],
                "first_5_values_red": img_input[0, 0, meta["pad"][1], meta["pad"][0]:meta["pad"][0] + 5].tolist()
            }, f, indent=2)
        # -----------------------------------------------------
        # 2. Run ONNX
        start = time.time()
        outputs = self.session.run(None, {'images': img_input})
        elapsed = (time.time() - start) * 1000

        preds = outputs[0][0]  # [300, 38]
        protos = outputs[1][0]  # [32, 160, 160]

        # 3. Confidence Filter
        mask = preds[:, 4] > conf
        preds = preds[mask]
        if len(preds) == 0: return [], 0, elapsed

        # 4. Class-Agnostic NMS (OpenCV Standalone)
        boxes_for_nms = []
        for b in preds[:, :4]:
            boxes_for_nms.append([float(b[0]), float(b[1]), float(b[2] - b[0]), float(b[3] - b[1])])

        indices = cv2.dnn.NMSBoxes(boxes_for_nms, preds[:, 4].tolist(), conf, iou)
        if len(indices) > 0:
            preds = preds[indices.flatten()]
        else:
            return [], 0, elapsed

        # 5. Rescale Boxes
        r = meta["ratio"]
        left, top = meta["pad"]
        boxes = preds[:, :4]
        boxes[:, [0, 2]] -= left
        boxes[:, [1, 3]] -= top
        boxes /= r

        # 6. Generate Tags and Masks
        tags = [self.vocab[int(c)] for c in preds[:, 5]]
        masks = self.process_masks(protos, preds[:, 6:], meta)

        return tags, len(masks), elapsed


def main():
    # Make sure you use the Dynamic ONNX exported previously
    MODEL_ONNX = "assets/model/yoloe-26n-seg-pf.onnx"
    VOCAB_JSON = "assets/model/vocabulary.json"
    IMG_DIR = Path("assets/img")

    predictor = YOLO26StandalonePredictor(MODEL_ONNX, VOCAB_JSON)

    print(f"--- YOLO26 Standalone Inference (No Ultralytics Dep) ---\n")
    for img_path in IMG_DIR.glob("*"):
        if img_path.suffix.lower() not in ['.jpg', '.png', '.jpeg', '.webp']: continue

        tags, mask_count, dt = predictor.predict(img_path)
        print(f"Image: {img_path.name} ({dt:.1f}ms)")
        print(f"  - Objects Found: {len(tags)}")
        if tags:
            print(f"  - Tags: {', '.join(set(tags))}")
        print("-" * 30)


if __name__ == "__main__":
    main()
